libdnf/rpm/solv/goal_private.cpp

```diff
@@ -632,8 +632,22 @@ transaction::TransactionItemReason GoalPrivate::get_reason(Id id) {
 
     if ((reason == SOLVER_REASON_UNIT_RULE || reason == SOLVER_REASON_RESOLVE_JOB) &&
         (solver_ruleclass(libsolv_solver, info) == SOLVER_RULE_JOB ||
-         solver_ruleclass(libsolv_solver, info) == SOLVER_RULE_BEST))
+         solver_ruleclass(libsolv_solver, info) == SOLVER_RULE_BEST)) {
+        // explicitely user-installed
+        for (const auto & ui_id : user_installed) {
```

Comment:
Sequential search is quite slow (in comparison to ordered list, set or map). For the transaction with a lot of package to be installed it can be measurable. Think about anaconda - there is a lot of packages installed by groups. The scan is also applied for removed|replaced packages and it is completely unneeded and may be harmful.